core/trino-main/src/main/java/io/trino/sql/analyzer/Analysis.java

```diff
@@ -242,6 +245,9 @@ public void setUpdateType(String updateType, QualifiedObjectName targetName, Opt
     {
         this.updateType = updateType;
         this.target = Optional.of(new UpdateTarget(targetName, targetTable, targetColumns));
+        if ("DELETE".equals(updateType) || "UPDATE".equals(updateType)) {
```

Comment:
This seems hacky and brittle.

I think the resetting of update type should be handled differently in the analyzer. Instead setting/resetting, the analysis of explain should be done with a context that indicates it's an explain query, so setUpdateType should not be called at all.

Reply:
If we don't call `setUpdateType` in case of `EXPLAIN` or `EXPLAIN ANALYZE` - how would we set the `TableScan#isUpdateNode` . which is based on `setUpdateType`. What if we could introduce another API in `Analysis` that captures the updatable table - which is invoked during `DELETE` or `UPDATE` operation - thereby we could remove the relationship between state of `updateType`.